WIP: Add functionality to track bitmap information and events

tamimattafi opened the pull request:
Hello! Thank you for this awesome library! 

We are strongly using it for our Story functionality such as on Instagram, and we are willing to help to modify, improve and add more functionality to this library.

Our use-case, that made us create this merge-request, includes the following points:
- Add the ability to attach buttons/links to a story
- Add the ability to edit the information of these attachments during creation
- Add the ability to recreate clickable areas on these attachments and do something if the user interacts

So, changes that we added to the library include these points:
- Made `ViewType` a `sealed interface` instead of `enum` and used the correct naming conventions for children.
- Added a new `ViewType`, called `Custom<T>`, that takes generic data as a parameter/property, this `ViewType` can be passed with the `bitmap` through `photoEditor.addImage(bitmap, viewType)`, if nothing is passed, `ViewType.Image` will be used by default.
- Changed `OnPhotoEditorListener` interface to return the `rootView` on each event, so we can extract the `Rect` and other information about the view and its properties
- `OnPhotoEditorListener.onAddViewListener` is now called through `rootView.post {}` to make sure that this event is fired only after the view finishes its creation and setup

For now, this is enough to extract view's data and track it, also to create clickable ghosts during story preview.

I hope this Pull Request will be accepted, it would be awesome to have a review first, so I could adjust any errors since I'm not that well-involved into this library and can miss some moments. Thank you!

burhanrashid52 commented on CHANGELOG.md:
```diff
@@ -84,4 +84,8 @@ new TextStyleBuilder()
\ No newline at end of file
+- New : Migrated the app and library to Kotlin
+
+### 2.1.0-alpha01
```
We don't have an alpha stage, so we can directly bump it to `2.1.0.`. This applies to all the files where we have used alpha versioning.

tamimattafi replied:
Fixed

burhanrashid52 commented on app/src/main/java/com/burhanrashid52/photoediting/EditImageActivity.kt:
```diff
@@ -183,20 +184,32 @@ class EditImageActivity : BaseActivity(), OnPhotoEditorListener, View.OnClickLis
-        Log.d(TAG, "onAddViewListener() called with: viewType = [$viewType], numberOfAddedViews = [$numberOfAddedViews]")
+    override fun onAddViewListener(rootView: View?, viewType: ViewType?, numberOfAddedViews: Int) {
+        val viewRect = Rect()
+        rootView?.getGlobalVisibleRect(viewRect)
```
Not sure what `getGlobalVisibleRect` does here, but it would be really helpful if you add an example of your use case here in a more general way like editing the view, adding links, and clicking links in this sample app.
This will also help us to understand how to us these new APIs.

burhanrashid52 replied:
For example, check `app/src/main/java/com/burhanrashid52/photoediting/EditImageActivity.java` in #358

tamimattafi replied:
`getGlobalVisibleRect` is used to get the position of the `rootView` on the screen, it is important since we need to save the exact coordinates of the view to recreate it (if we removed it before saving) or to add a clickable ghost above it. A use-case could be a story like instagram, with a clickable tag or link or location.

I can add such use-case to the sample but it would take some time, maybe for a future collaboration:)

burhanrashid52 commented on photoeditor/src/main/java/ja/burhanrashid52/photoeditor/MultiTouchListener.kt:
```diff
@@ -111,8 +111,12 @@ internal class MultiTouchListener(
         if (mOnPhotoEditorListener != null && viewTag != null && viewTag is ViewType) {
-            if (isStart) mOnPhotoEditorListener.onStartViewChangeListener(view.tag as ViewType) else mOnPhotoEditorListener.onStopViewChangeListener(
-                view.tag as ViewType
+            if (isStart) mOnPhotoEditorListener.onStartViewChangeListener(
```
Nitpick: Adding parentheses.

tamimattafi replied:
Fixed

burhanrashid52 commented on photoeditor/src/main/java/ja/burhanrashid52/photoeditor/ViewType.kt:
```diff
@@ -10,9 +10,10 @@ package ja.burhanrashid52.photoeditor
-    IMAGE,
-    EMOJI
+sealed interface ViewType {
+    object BrushDrawer : ViewType
```
Better to rename from `BrushDrawer` instead of `Brush`

tamimattafi replied:
Fixed

burhanrashid52 reviewed (changes requested):
Thank you for this PR. I think this will open up a lot of customization options for image editing.

However, I am curious to know how clicking implementation will work, since this library saves a static image?

burhanrashid52 commented:
@tamimattafi Also, please make sure all the test passes. There are some failing tests on the current PR.

burhanrashid52 commented:
@tamimattafi Are there any updates on this?

tamimattafi commented:
Hello @burhanrashid52,
I apologize for being inactive and not responding to your comments. I was quite busy and didn't really check my notifications.
Thank you for your review, I can fix the issues and add some use-case examples if we can re-open the pull request

burhanrashid52 commented:
> Hello @burhanrashid52, I apologize for being inactive and not responding to your comments. I was quite busy and didn't really check my notifications. Thank you for your review, I can fix the issues and add some use-case examples if we can re-open the pull request

Sure. I am reopening it.

tamimattafi commented:
> > Hello @burhanrashid52, I apologize for being inactive and not responding to your comments. I was quite busy and didn't really check my notifications. Thank you for your review, I can fix the issues and add some use-case examples if we can re-open the pull request
> 
> Sure. I am reopening it.

Thank you, I will make sure to do it this weekend.

tamimattafi commented:
Hello @burhanrashid52
I have checked my local code, I've made some changes before your review, fixed some issues and refactored some code. Should I close the Pull Request and open a new one, or I should update the current on with the new changes + add the changes you requested?

burhanrashid52 commented:
> Hello @burhanrashid52 I have checked my local code, I've made some changes before your review, fixed some issues and refactored some code. Should I close the Pull Request and open a new one, or I should update the current on with the new changes + add the changes you requested?

If you've changed the majority of the structure then better to open a new PR and if you just address changes regarding my comment then it's better to keep the PR itself to track reviews.

burhanrashid52 commented:
@tamimattafi I am planning to publish a new version [3.0.1](https://github.com/burhanrashid52/PhotoEditor/issues?q=label%3A3.0.1). So wanted to check up to see when can we expect the new PR. Or else I publish the new version for now.

tamimattafi commented:
@burhanrashid52 Sorry again for the delay, we have some troubles at work. I think you should publish the new release for now without my changes, I will try to prepare a new PR this weekend (Probably Sunday)

tamimattafi commented on CHANGELOG.md:
```diff
@@ -84,4 +84,8 @@ new TextStyleBuilder()
+
+### 2.1.0-alpha01
+- New: Add custom ViewType
+- New: Add rootView to OnPhotoEditorListener events
```
Fixed
